clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -373,34 +374,112 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 
   exchange(uint8_t *local_memory) : _local_memory(local_memory) {}
 
-  /// Rearrange elements from rank order to blocked order
-  template <typename Item>
-  __dpct_inline__ void
-  scatter_to_blocked(Item item, T (&keys)[VALUES_PER_THREAD],
-                     int (&ranks)[VALUES_PER_THREAD]) {
+  // TODO: Investigate if padding is required for performance, 
+  // and if specializations are required for specific target hardware.
+  static size_t adjust_by_padding(size_t offset) {
+
+    if constexpr (INSERT_PADDING) {
+      offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    }
+    return offset;
+  }
+
+  template <typename Item> struct getStripedOffset {
+    size_t operator()(Item item, int i) {
+      size_t offset =
+          item.get_local_id(0) * VALUES_PER_THREAD + i;
+      return adjust_by_padding(offset);
+    }
+  };
```

Comment:
```suggestion
  struct getStripedOffset {
    template <typename Item>
    size_t operator()(Item item, int i) {
      size_t offset =
          item.get_local_id(0) * VALUES_PER_THREAD + i;
      return adjust_by_padding(offset);
    }
  };
```

If you make the `operator()` the template rather than the `struct`, then you don't have to specify the type for all the operators when you create them.  The operator itself will deduce the type required.

`getStripedOffset<Item> get_striped_offset;`
->
`getStripedOffset get_striped_offset;`

The same goes for the other two functors.